napari/_vispy/vispy_transforms.py

```diff
@@ -0,0 +1,136 @@
+        Transform chain for an ordered sequence of napari transforms.
+
+    Attributes
+    ----------
+    scale : ndarray, shape(4)
+        Combined scale of all chained transforms, vispy format (X,Y,Z,T).
+    simplified : vispy.visuals.transforms.ChainTransform
+        Simplified ChainTransform where transforms have been combined into one.
+        Used to reduce compute time, since a single matrix operation is faster.
+    transform_chain : vispy.visuals.transforms.ChainTransform
+        Transform chain for an ordered sequence of vispy transforms.
+    translate : ndarray, shape(4)
+        Combined tranlation of all chained transforms, vispy format (X,Y,Z,T).
+    """
+
+    def __init__(self, napari_transform_chain=TransformChain()):
+        self.napari_transform_chain = napari_transform_chain
+
+        self.transform_chain = self._vispy_transform_chain()
+        self._calculate_attributes()
+
+        self.changed = self.transform_chain.changed
+
+        self.napari_transform_chain.events.added.connect(self._add)
+        self.napari_transform_chain.events.removed.connect(self._remove)
+        self.napari_transform_chain.events.reordered.connect(self._reorder)
+
+    @property
+    def scale(self):
+        """Combined scale of all elements in the VispyChainTransform.
+
+        You should not directly modify the scale property of a
+        VispyChainTransform. Instead you should append or insert another
+        transform to the napari TransformChain, which will update this
+        property using event emitters.
+
+        Returns
+        -------
+        ndarray, shape (4,)
+            Combined scale of all elements in the VispyChainTransform.
+            Vispy format expects (X,Y,Z,T)
+        """
+        self._calculate_attributes()
+        return self._scale
+
+    @property
+    def translate(self):
+        """Combined translation of all elements in the VispyChainTransform.
+
+        You should not directly modify the translate property of a
+        VispyChainTransform. Instead you should append or insert another
+        transform to the napari TransformChain, which will update this
+        property using event emitters.
+
+        Returns
+        -------
+        ndarray, shape (4,)
+            Combined translation of all elements in the VispyChainTransform.
+            Vispy format expects (X,Y,Z,T)
+        """
+        self._calculate_attributes()
+        return self._translate
+
+    @property
+    def simplified(self):
+        return self._simplified
+
+    def _vispy_transform_chain(self):
+        """Builds vispy ChainTransform from a napari TransformChain."""
+        vispy_transforms = []
+        for t in self.napari_transform_chain:
+            # From napari format (t, z, y, x)
+            # to the vispy format (x, y, z, ?)
+            _translate = np.flip(t.translate)  # vispy ordered array
+            _scale = np.flip(t.scale)  # vispy ordered array
+            vispy_transforms.append(
```

Comment:
Is my understanding here correct? I think the napari transforms have the format (T,Z,Y,X), and the vispy transforms have the format (X,Y,Z,T). Additionally napari transform scale and translate arrays can have fewer than 4 elements, eg if you're just specifying a translation in x and y.

Reply:
I think napari transforms will always have N elements for an nD layer, and vispy transforms will always have the same number of elements because they just relate to changing what is changing on the screen. I need to look more into the vispy transforms again though to remember exactly how they are specified, but the order of the axes is reversed in napari / vispy as you note above

Reply:
It's not really clear to me what the final element of the vispy 4 vector does - but i think they call it W, see here https://github.com/vispy/vispy/blob/a6ec64bb7b56739b463e3c16245a31940ca4f5f7/vispy/visuals/transforms/linear.py#L54-L75 so we might as well put `(X, Y, Z, W)` in our docs, and I think we can try setting it to always be 1 or 0. So if you have a 5D translation like `(T, C, Z, Y, X)`, you'd end up with a vispy 4-vec of `(X, Y, Z, 1)` or `(X, Y, Z, 0)`

Reply:
It's not clear to me either, I guessed it might be something to do with a time axis but I don't think that's right.

When I stepped through the `find_top_left()` function with the debugger, the last element of the vispy 4 vectors was one, not zero. That might only be true in that particular context though. In any case, I think we always chop off the last element of that vector before using it for anything in napari.

Reply:
I think/presume that the fourth element relates to homogeneous coordinates:

https://en.wikipedia.org/wiki/Homogeneous_coordinates

Reply:
And yes, it should always be 1, if you look at that page.

Reply:
Yes, I was thinking a 1 for homogenous coordinates, but wasn't sure